### PR TITLE
Run|Debug hover actions.

### PR DESCRIPTION
**Pull request by vsrs**
hover_actions_run

This hover actions work exactly like corresponding lenses.

### CODE REVIEW

File: crates/rust-analyzer/src/main_loop/handlers.rs [hunk 1]
```diff
@@ -1169,6 +1134,22 @@ fn show_references_command(
...
 
+fn run_single_command(runnable: &lsp_ext::Runnable, title: &str) -> Command {
+    Command {
+        title: title.to_string(),
+        command: "rust-analyzer.runSingle".into(),
+        arguments: Some(vec![to_value(runnable).unwrap()]),
+    }
+}
+
+fn debug_single_command(runnable: lsp_ext::Runnable) -> Command {
```
Comment (matklad) [inline crates/rust-analyzer/src/main_loop/handlers.rs:1145]:
```suggestion
fn debug_single_command(runnable: &lsp_ext::Runnable) -> Command {
```

File: crates/rust-analyzer/src/to_proto.rs [hunk 2]
```diff
@@ -656,22 +656,14 @@ pub(crate) fn resolved_code_action(
 pub(crate) fn runnable(
     snap: &GlobalStateSnapshot,
     file_id: FileId,
-    runnable: Runnable,
+    runnable: &Runnable,
```
Comment (matklad) [inline crates/rust-analyzer/src/to_proto.rs:659]: I *think* semantically this should be `runnable: Runnable`. It's nice to statically enforce that we don't duplicate runnables. The lifetimes work our if you make `HoverResult::actions` public, which seems like a good idea -- so far it seems like ide-layer API work best if they are just plain data. 

It might make sense to make `results` public as well, but I'd say that they shold be wholesale replaced with something more specific (though this obviously should be handled by an unrelated PR):

```rust
struct HoverResult {
  ty: Option<Stirng>,
  docs: Option<String>,
  module: Option<String>,
}
```

File: editors/code/package.json [hunk 3]
```diff
@@ -486,6 +486,16 @@
...
+                "rust-analyzer.hoverActions.run": {
+                    "markdownDescription": "Whether to show `Run` action. Only applies when `#rust-analyzer.hoverActions.enable#` is set.",
+                    "type": "boolean",
+                    "default": true
+                },
+                "rust-analyzer.hoverActions.debug": {
+                    "markdownDescription": "Whether to show `Debug` action. Only applies when `#rust-analyzer.hoverActions.enable#` is set.",
+                    "type": "boolean",
+                    "default": true
+                },
```
Comment (matklad) [inline editors/code/package.json:498]: :+1: on adding fine-grained feature flags from the start!

File: crates/rust-analyzer/src/to_proto.rs (hunk 2 above)
Comment (vsrs) [reply crates/rust-analyzer/src/to_proto.rs:659]: Might be on this level (ra_ide::HoverResult) we can even have not only a string, but something more specific (Type, Function, Local, etc.)?

Comment (matklad) [reply crates/rust-analyzer/src/to_proto.rs]: Heh, looks like the strategy of documenting knowledge works :)

https://github.com/rust-analyzer/rust-analyzer/pull/4792/commits/ee8dec5dc11cfecf219b6510b0eadd9691a82ba5

TL;DR: eventually, we *have* to show the info in the editor, where compiler concepts like definition and usages do not work (but, eg, strings and icons do). The `ra_ide` is exactly the layer where we switch from "compiler view" to the "editor" view, so in general ide speaks in generalized `Display` things.

**Comment by matklad** [comment]
@vsrs could you remove @mention from the commit message? They create pings when folks run rebases in their forks :)

File: crates/rust-analyzer/src/to_proto.rs (hunk 2 above)
Comment (vsrs) [reply crates/rust-analyzer/src/to_proto.rs]: it does :)

Got it. Thanks.

**Comment by matklad** [comment]
> They create pings when folks run rebases in their forks :)

Like this :)

image

image

**Comment by vsrs** [comment]
> @vsrs could you remove @mention from the commit message? They create pings when folks run rebases in their forks :)

Ups, sorry. Sure, I'l remove it.

**Comment by matklad** [comment]
I assume 

bors r+

?

**Comment by vsrs** [comment]
Yes. Didn't notice that all checks were already completed.

**Comment by matklad** [comment]
>Yes. Didn't notice that all checks were already completed.

You don't need to wait until the checks are completed. It's totally fine to r+ immediately after the push, the PR won't be merged if it fails the checks. Generally, the earlier you r+, the better.

**Comment by vsrs** [comment]
> Generally, the earlier you r+, the better.
OK. I'll remember that next time.
